### PR TITLE
gh-117953: Always Run Extension Init Func in Main Interpreter First

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
This change makes sure all extension/builtin modules have their init function run first by the main interpreter before proceeding with import in the original interpreter (main or otherwise).  This means when the import of a single-phase init module fails in an isolated subinterpreter, it won't tie any global state/callbacks to the subinterpreter.

This supercedes gh-117487.

* Issue: gh-117953

### CODE REVIEW

**Comment by ericsnowcurrently** [comment]
FYI, there's a refleak somewhere in here that I'm tracking down.

File: Python/import.c [hunk 1]
```diff
@@ -1874,38 +1894,82 @@ import_run_extension(PyThreadState *tstate, PyModInitFunction p0,
...
+     * To avoid all of that, we make sure the module's init function
+     * is always run first with the main interpreter active.  If it was
+     * already the main interpreter then we can continue loading the
+     * module like normal.  Otherwise, right after the init function,
+     * we take care of some import state bookkeeping, switch back
+     * to the subinterpreter, check for single-phase init,
+     * and then continue loading like normal. */
+
+    PyThreadState *main_tstate = NULL;
+    if (!_Py_IsMainInterpreter(tstate->interp)) {
```
Comment (zooba) [inline Python/import.c]: Is this necessary if `switch_to_main_interpreter` is doing the same check?

File: Python/import.c [hunk 2]
```diff
@@ -1941,15 +2005,73 @@ import_run_extension(PyThreadState *tstate, PyModInitFunction p0,
         if (cached == NULL) {
             goto error;
```
Comment (zooba) [inline Python/import.c]: Can we do this without switching back?

**Review by zooba — COMMENTED** [review]
Nothing jumps out at me looking like a refleak, unfortunately. Possible the first import is leaking something in the main interpreter and it's not being counted right in the subinterpreter? (Or vice versa)

File: Python/import.c (hunk 2 above)
Comment (ericsnowcurrently) [reply Python/import.c]: Yeah, I noticed that too and have already fixed it locally. 😄

File: Python/import.c (hunk 1 above)
Comment (ericsnowcurrently) [reply Python/import.c]: It isn't necessary, but I wanted to be explicit about it.  I'll see about cleaning that up.

Comment (ericsnowcurrently) [reply Python/import.c]: fixed

File: Python/import.c (hunk 2 above)
Comment (ericsnowcurrently) [reply Python/import.c]: fixed

**Pull request merged**